From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Ser <contact@emersion.fr>
Date: Thu, 9 Oct 2025 17:38:23 +0200
Subject: [PATCH 3/8] cta: introduce struct di_cta

In the next commit, CTA parsing functions will be re-used for
DisplayID v2. However, DisplayID v2 doesn't embed a full EDID-style
CTA extension. Instead, it embeds a list of CTA blocks.

Stop passing around a struct di_edid_cta and instead pass a
struct di_cta with minimal state.

Signed-off-by: Simon Ser <contact@emersion.fr>
(cherry picked from commit d535192adc496335b9ff2268398709e4cd566e68)
---
SPDX-License-Identifier: MIT
Upstream: https://gitlab.freedesktop.org/emersion/libdisplay-info/-/commit/d535192adc496335b9ff2268398709e4cd566e68
Backported onto libdisplay-info 0.3.0 for ally-x20-steamos (soname kept at
libdisplay-info.so.3; test/data hunks dropped). The MIT license and
copyright notice of libdisplay-info apply; see LICENSE.md.

 cta.c         | 120 +++++++++++++++++++++++++++-----------------------
 include/cta.h |  12 ++++-
 2 files changed, 76 insertions(+), 56 deletions(-)

diff --git a/cta.c b/cta.c
index 53edfca..dec27e6 100644
--- a/cta.c
+++ b/cta.c
@@ -59,7 +59,7 @@ di_cta_hdmi_video_format_from_hdmi_vic(uint8_t hdmi_vic)
 }
 
 static void
-add_failure(struct di_edid_cta *cta, const char fmt[], ...)
+add_failure(struct di_cta *cta, const char fmt[], ...)
 {
 	va_list args;
 
@@ -69,7 +69,7 @@ add_failure(struct di_edid_cta *cta, const char fmt[], ...)
 }
 
 static void
-add_failure_until(struct di_edid_cta *cta, int revision, const char fmt[], ...)
+add_failure_until(struct di_cta *cta, int revision, const char fmt[], ...)
 {
 	va_list args;
 
@@ -83,7 +83,7 @@ add_failure_until(struct di_edid_cta *cta, int revision, const char fmt[], ...)
 }
 
 static struct di_cta_svd *
-parse_svd(struct di_edid_cta *cta, uint8_t raw, uint8_t original_index,
+parse_svd(struct di_cta *cta, uint8_t raw, uint8_t original_index,
 	  const char *prefix)
 {
 	struct di_cta_svd svd, *svd_ptr;
@@ -116,7 +116,7 @@ parse_svd(struct di_edid_cta *cta, uint8_t raw, uint8_t original_index,
 }
 
 static bool
-parse_video_block(struct di_edid_cta *cta, struct di_cta_video_block_priv *video,
+parse_video_block(struct di_cta *cta, struct di_cta_video_block_priv *video,
 		  const uint8_t *data, size_t size)
 {
 	uint8_t i;
@@ -140,7 +140,7 @@ parse_video_block(struct di_edid_cta *cta, struct di_cta_video_block_priv *video
 }
 
 static int
-hdmi_latency_from_raw(struct di_edid_cta *cta, const char *block_name,
+hdmi_latency_from_raw(struct di_cta *cta, const char *block_name,
 		      const char *type, uint8_t raw)
 {
 	/* Unknown latency.  */
@@ -162,7 +162,7 @@ hdmi_latency_from_raw(struct di_edid_cta *cta, const char *block_name,
 }
 
 static bool
-parse_vendor_hdmi_block(struct di_edid_cta *cta,
+parse_vendor_hdmi_block(struct di_cta *cta,
 			struct di_cta_vendor_hdmi_block_priv *priv,
 			const uint8_t *data, size_t size)
 {
@@ -310,7 +310,7 @@ parse_vendor_hdmi_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_hdmi_scds(struct di_edid_cta *cta, struct di_cta_hdmi_scds *scds,
+parse_hdmi_scds(struct di_cta *cta, struct di_cta_hdmi_scds *scds,
 		struct di_cta_hdmi_dsc *dsc, const uint8_t *data, size_t size,
 		const char *block_name)
 {
@@ -477,7 +477,7 @@ parse_hdmi_scds(struct di_edid_cta *cta, struct di_cta_hdmi_scds *scds,
 }
 
 static bool
-parse_vendor_hdmi_forum_block(struct di_edid_cta *cta,
+parse_vendor_hdmi_forum_block(struct di_cta *cta,
 			      struct di_cta_vendor_hdmi_forum_block_priv *priv,
 			      const uint8_t *data, size_t size)
 {
@@ -492,7 +492,7 @@ parse_vendor_hdmi_forum_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_hdmi_forum_sink_cap(struct di_edid_cta *cta,
+parse_hdmi_forum_sink_cap(struct di_cta *cta,
 			  struct di_cta_hdmi_forum_sink_cap_priv *priv,
 			  const uint8_t *data, size_t size)
 {
@@ -507,7 +507,7 @@ parse_hdmi_forum_sink_cap(struct di_edid_cta *cta,
 }
 
 static bool
-parse_ycbcr420_block(struct di_edid_cta *cta,
+parse_ycbcr420_block(struct di_cta *cta,
 		     struct di_cta_ycbcr420_video_block_priv *ycbcr420,
 		     const uint8_t *data, size_t size)
 {
@@ -532,7 +532,7 @@ parse_ycbcr420_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_sad_format(struct di_edid_cta *cta, uint8_t code, uint8_t code_ext,
+parse_sad_format(struct di_cta *cta, uint8_t code, uint8_t code_ext,
 		 enum di_cta_audio_format *format, const char *prefix)
 {
 	switch (code) {
@@ -625,7 +625,7 @@ parse_sad_format(struct di_edid_cta *cta, uint8_t code, uint8_t code_ext,
 }
 
 static bool
-parse_sad(struct di_edid_cta *cta, struct di_cta_audio_block_priv *audio,
+parse_sad(struct di_cta *cta, struct di_cta_audio_block_priv *audio,
 	  const uint8_t data[static CTA_SAD_SIZE])
 {
 	enum di_cta_audio_format format;
@@ -902,7 +902,7 @@ parse_sad(struct di_edid_cta *cta, struct di_cta_audio_block_priv *audio,
 }
 
 static bool
-parse_audio_block(struct di_edid_cta *cta, struct di_cta_audio_block_priv *audio,
+parse_audio_block(struct di_cta *cta, struct di_cta_audio_block_priv *audio,
 		  const uint8_t *data, size_t size)
 {
 	size_t i;
@@ -920,7 +920,7 @@ parse_audio_block(struct di_edid_cta *cta, struct di_cta_audio_block_priv *audio
 }
 
 static bool
-parse_speaker_alloc(struct di_edid_cta *cta, struct di_cta_speaker_allocation *speaker_alloc,
+parse_speaker_alloc(struct di_cta *cta, struct di_cta_speaker_allocation *speaker_alloc,
 		    const uint8_t data[3], const char *prefix)
 {
 	bool rlc_rrc;
@@ -961,7 +961,7 @@ parse_speaker_alloc(struct di_edid_cta *cta, struct di_cta_speaker_allocation *s
 }
 
 static bool
-parse_speaker_alloc_block(struct di_edid_cta *cta,
+parse_speaker_alloc_block(struct di_cta *cta,
 			  struct di_cta_speaker_alloc_block *speaker_alloc,
 			  const uint8_t *data, size_t size)
 {
@@ -979,7 +979,7 @@ parse_speaker_alloc_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_video_cap_block(struct di_edid_cta *cta,
+parse_video_cap_block(struct di_cta *cta,
 		      struct di_cta_video_cap_block *video_cap,
 		      const uint8_t *data, size_t size)
 {
@@ -1004,11 +1004,11 @@ parse_video_cap_block(struct di_edid_cta *cta,
 
 	switch (video_cap->it_over_underscan) {
 	case DI_CTA_VIDEO_CAP_ALWAYS_OVERSCAN:
-		if (cta->flags.it_underscan)
+		if (cta->flags->it_underscan)
 			add_failure(cta, "Video Capability Data Block: IT video formats are always overscanned, but bit 7 of Byte 3 of the CTA-861 Extension header is set to underscanned.");
 		break;
 	case DI_CTA_VIDEO_CAP_ALWAYS_UNDERSCAN:
-		if (!cta->flags.it_underscan)
+		if (!cta->flags->it_underscan)
 			add_failure(cta, "Video Capability Data Block: IT video formats are always underscanned, but bit 7 of Byte 3 of the CTA-861 Extension header is set to overscanned.");
 	default:
 		break;
@@ -1069,7 +1069,7 @@ parse_vesa_display_device_additional_primary_chromaticity(struct di_cta_vesa_dis
 }
 
 static bool
-parse_vesa_display_device(struct di_edid_cta *cta, struct di_cta_vesa_display_device_block *dddb,
+parse_vesa_display_device(struct di_cta *cta, struct di_cta_vesa_display_device_block *dddb,
 		const uint8_t *data, size_t size)
 {
 	const size_t offset = 2; /* CTA block header */
@@ -1271,7 +1271,7 @@ parse_vesa_display_device(struct di_edid_cta *cta, struct di_cta_vesa_display_de
 }
 
 static bool
-parse_colorimetry_block(struct di_edid_cta *cta,
+parse_colorimetry_block(struct di_cta *cta,
 			struct di_cta_colorimetry_block *colorimetry,
 			const uint8_t *data, size_t size)
 {
@@ -1317,7 +1317,7 @@ parse_min_luminance(uint8_t raw, float max)
 }
 
 static bool
-parse_hdr_static_metadata_block(struct di_edid_cta *cta,
+parse_hdr_static_metadata_block(struct di_cta *cta,
 				struct di_cta_hdr_static_metadata_block_priv *metadata,
 				const uint8_t *data, size_t size)
 {
@@ -1360,7 +1360,7 @@ parse_hdr_static_metadata_block(struct di_edid_cta *cta,
 	return true;
 }
 static bool
-parse_hdr_dynamic_metadata_block(struct di_edid_cta *cta,
+parse_hdr_dynamic_metadata_block(struct di_cta *cta,
 				 struct di_cta_hdr_dynamic_metadata_block_priv *priv,
 				 const uint8_t *data, size_t size)
 {
@@ -1474,7 +1474,7 @@ parse_hdr_dynamic_metadata_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_vesa_transfer_characteristics_block(struct di_edid_cta *cta,
+parse_vesa_transfer_characteristics_block(struct di_cta *cta,
 					  struct di_cta_vesa_transfer_characteristics_block *tf,
 					  const uint8_t *data, size_t size)
 {
@@ -1497,7 +1497,7 @@ parse_vesa_transfer_characteristics_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_video_format_pref_block(struct di_edid_cta *cta,
+parse_video_format_pref_block(struct di_cta *cta,
 			      struct di_cta_video_format_pref_priv *vfpdb,
 			      const uint8_t *data, size_t size)
 {
@@ -1548,7 +1548,7 @@ parse_video_format_pref_block(struct di_edid_cta *cta,
 
 
 static void
-parse_ycbcr420_cap_map(struct di_edid_cta *cta,
+parse_ycbcr420_cap_map(struct di_cta *cta,
 		       struct di_cta_ycbcr420_cap_map_block *ycbcr420_cap_map,
 		       const uint8_t *data, size_t size)
 {
@@ -1562,7 +1562,7 @@ parse_ycbcr420_cap_map(struct di_edid_cta *cta,
 }
 
 static bool
-parse_hdmi_audio_3d_descriptor(struct di_edid_cta *cta,
+parse_hdmi_audio_3d_descriptor(struct di_cta *cta,
 			       struct di_cta_sad_priv *sad,
 			       const uint8_t *data, size_t size)
 {
@@ -1611,7 +1611,7 @@ parse_hdmi_audio_3d_descriptor(struct di_edid_cta *cta,
 }
 
 static bool
-parse_hdmi_audio_block(struct di_edid_cta *cta,
+parse_hdmi_audio_block(struct di_cta *cta,
 		       struct di_cta_hdmi_audio_block_priv *priv,
 		       const uint8_t *data, size_t size)
 {
@@ -1705,7 +1705,7 @@ skip:
 }
 
 static struct di_cta_infoframe_descriptor *
-parse_infoframe(struct di_edid_cta *cta, uint8_t type,
+parse_infoframe(struct di_cta *cta, uint8_t type,
 		const uint8_t *data, size_t size)
 {
 	struct di_cta_infoframe_descriptor infoframe = {0};
@@ -1760,7 +1760,7 @@ parse_infoframe(struct di_edid_cta *cta, uint8_t type,
 }
 
 static bool
-parse_infoframe_block(struct di_edid_cta *cta,
+parse_infoframe_block(struct di_cta *cta,
 		      struct di_cta_infoframe_block_priv *ifb,
 		      const uint8_t *data, size_t size)
 {
@@ -1828,7 +1828,7 @@ decode_coord(unsigned char x)
 }
 
 static bool
-parse_room_config_block(struct di_edid_cta *cta,
+parse_room_config_block(struct di_cta *cta,
 			struct di_cta_room_configuration_block *rc,
 			const uint8_t *data, size_t size)
 {
@@ -1895,7 +1895,7 @@ parse_room_config_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_speaker_location_block(struct di_edid_cta *cta,
+parse_speaker_location_block(struct di_cta *cta,
 			     struct di_cta_speaker_location_priv *sldb,
 			     const uint8_t *data, size_t size)
 {
@@ -1947,7 +1947,7 @@ parse_speaker_location_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_did_type_vii_timing(struct di_edid_cta *cta,
+parse_did_type_vii_timing(struct di_cta *cta,
 			  struct di_cta_type_vii_timing_priv *t,
 			  const uint8_t *data, size_t size)
 {
@@ -2068,7 +2068,7 @@ ff_peak_lum_index_to_nits(int index, int peak_lum)
 }
 
 static bool
-parse_hdr10plus_block(struct di_edid_cta *cta,
+parse_hdr10plus_block(struct di_cta *cta,
 		      struct di_cta_hdr10plus_block *block,
 		      const uint8_t *data, size_t size)
 {
@@ -2106,7 +2106,7 @@ parse_hdr10plus_block(struct di_edid_cta *cta,
 }
 
 static bool
-parse_dolby_video_block(struct di_edid_cta *cta,
+parse_dolby_video_block(struct di_cta *cta,
 			 struct di_cta_dolby_video_block_priv *dv_priv,
 			 const uint8_t *data, size_t size)
 {
@@ -2357,7 +2357,7 @@ destroy_data_block(struct di_cta_data_block *data_block)
 }
 
 static bool
-parse_vendor_specific_video_block(struct di_edid_cta *cta,
+parse_vendor_specific_video_block(struct di_cta *cta,
 				  enum di_cta_data_block_tag *tag,
 				  struct di_cta_data_block *data_block,
 				  const uint8_t *data, size_t size)
@@ -2400,7 +2400,7 @@ skip:
 }
 
 static bool
-parse_vendor_specific_block(struct di_edid_cta *cta,
+parse_vendor_specific_block(struct di_cta *cta,
 			    enum di_cta_data_block_tag *tag,
 			    struct di_cta_data_block *data_block,
 			    const uint8_t *data, size_t size)
@@ -2440,12 +2440,15 @@ skip:
 }
 
 static bool
-parse_data_block(struct di_edid_cta *cta, uint8_t raw_tag, const uint8_t *data, size_t size)
+parse_data_block(struct di_cta *cta, uint8_t raw_tag, const uint8_t *data, size_t size,
+		 struct di_cta_data_block **data_block_out)
 {
 	enum di_cta_data_block_tag tag;
 	uint8_t extended_tag;
 	struct di_cta_data_block *data_block;
 
+	*data_block_out = NULL;
+
 	data_block = calloc(1, sizeof(*data_block));
 	if (!data_block) {
 		return false;
@@ -2630,8 +2633,7 @@ parse_data_block(struct di_edid_cta *cta, uint8_t raw_tag, const uint8_t *data,
 	}
 
 	data_block->tag = tag;
-	assert(cta->data_blocks_len < EDID_CTA_MAX_DATA_BLOCKS);
-	cta->data_blocks[cta->data_blocks_len++] = data_block;
+	*data_block_out = data_block;
 	return true;
 
 skip:
@@ -2644,29 +2646,35 @@ error:
 }
 
 bool
-_di_edid_cta_parse(struct di_edid_cta *cta, const uint8_t *data, size_t size,
+_di_edid_cta_parse(struct di_edid_cta *edid_cta, const uint8_t *data, size_t size,
 		   struct di_logger *logger)
 {
+	struct di_cta *cta;
 	uint8_t flags, dtd_start;
 	uint8_t data_block_header, data_block_tag, data_block_size;
 	size_t i;
 	struct di_edid_detailed_timing_def_priv *detailed_timing_def;
+	struct di_cta_data_block *data_block;
 
 	assert(size == 128);
 	assert(data[0] == 0x02);
 
-	cta->logger = logger;
-
-	cta->revision = data[1];
+	edid_cta->revision = data[1];
 	dtd_start = data[2];
 
+	cta = &(struct di_cta){
+		.revision = edid_cta->revision,
+		.flags = &edid_cta->flags,
+		.logger = logger,
+	};
+
 	flags = data[3];
 	if (cta->revision >= 2) {
-		cta->flags.it_underscan = has_bit(flags, 7);
-		cta->flags.basic_audio = has_bit(flags, 6);
-		cta->flags.ycc444 = has_bit(flags, 5);
-		cta->flags.ycc422 = has_bit(flags, 4);
-		cta->flags.native_dtds = get_bit_range(flags, 3, 0);
+		edid_cta->flags.it_underscan = has_bit(flags, 7);
+		edid_cta->flags.basic_audio = has_bit(flags, 6);
+		edid_cta->flags.ycc444 = has_bit(flags, 5);
+		edid_cta->flags.ycc422 = has_bit(flags, 4);
+		edid_cta->flags.native_dtds = get_bit_range(flags, 3, 0);
 	} else if (flags != 0) {
 		/* Reserved */
 		add_failure(cta, "Non-zero byte 3.");
@@ -2698,11 +2706,16 @@ _di_edid_cta_parse(struct di_edid_cta *cta, const uint8_t *data, size_t size,
 		}
 
 		if (!parse_data_block(cta, data_block_tag,
-				      &data[i + 1], data_block_size)) {
-			_di_edid_cta_finish(cta);
+				      &data[i + 1], data_block_size, &data_block)) {
+			_di_edid_cta_finish(edid_cta);
 			return false;
 		}
 
+		if (data_block != NULL) {
+			assert(edid_cta->data_blocks_len < EDID_CTA_MAX_DATA_BLOCKS);
+			edid_cta->data_blocks[edid_cta->data_blocks_len++] = data_block;
+		}
+
 		i += 1 + data_block_size;
 	}
 
@@ -2718,11 +2731,11 @@ _di_edid_cta_parse(struct di_edid_cta *cta, const uint8_t *data, size_t size,
 
 		detailed_timing_def = _di_edid_parse_detailed_timing_def(&data[i]);
 		if (!detailed_timing_def) {
-			_di_edid_cta_finish(cta);
+			_di_edid_cta_finish(edid_cta);
 			return false;
 		}
-		assert(cta->detailed_timing_defs_len < EDID_CTA_MAX_DETAILED_TIMING_DEFS);
-		cta->detailed_timing_defs[cta->detailed_timing_defs_len++] = detailed_timing_def;
+		assert(edid_cta->detailed_timing_defs_len < EDID_CTA_MAX_DETAILED_TIMING_DEFS);
+		edid_cta->detailed_timing_defs[edid_cta->detailed_timing_defs_len++] = detailed_timing_def;
 	}
 
 	/* All padding bytes after the last DTD must be zero */
@@ -2734,7 +2747,6 @@ _di_edid_cta_parse(struct di_edid_cta *cta, const uint8_t *data, size_t size,
 		i++;
 	}
 
-	cta->logger = NULL;
 	return true;
 }
 
diff --git a/include/cta.h b/include/cta.h
index 25e1dfa..9405932 100644
--- a/include/cta.h
+++ b/include/cta.h
@@ -81,6 +81,16 @@
  */
 #define EDID_CTA_MAX_HDMI_AUDIO_BLOCK_ENTRIES 15
 
+/**
+ * Context for CTA parsing functions.
+ */
+struct di_cta {
+	int revision;
+	const struct di_edid_cta_flags *flags;
+
+	struct di_logger *logger;
+};
+
 struct di_edid_cta {
 	int revision;
 	struct di_edid_cta_flags flags;
@@ -92,8 +102,6 @@ struct di_edid_cta {
 	/* NULL-terminated */
 	struct di_edid_detailed_timing_def_priv *detailed_timing_defs[EDID_CTA_MAX_DETAILED_TIMING_DEFS + 1];
 	size_t detailed_timing_defs_len;
-
-	struct di_logger *logger;
 };
 
 struct di_cta_hdr_static_metadata_block_priv {
